src/core/Oscillator.cpp

```diff
@@ -24,72 +25,288 @@
 
 #include "Oscillator.h"
 
+#include <algorithm>
+#if !defined(__MINGW32__) && !defined(__MINGW64__)
+	#include <thread>
+#endif
+
 #include "BufferManager.h"
 #include "Engine.h"
 #include "Mixer.h"
 #include "AutomatableModel.h"
+#include "fftw3.h"
+#include "fft_helpers.h"
+
 
 
+void Oscillator::waveTableInit()
+{
+	createFFTPlans();
+	generateWaveTables();
+	// The oscillator FFT plans remain throughout the application lifecycle
+	// due to being expensive to create, and being used whenever a userwave form is changed
+	// deleted in main.cpp main()
+
+}
 
-Oscillator::Oscillator( const IntModel * _wave_shape_model,
-				const IntModel * _mod_algo_model,
-				const float & _freq,
-				const float & _detuning,
-				const float & _phase_offset,
-				const float & _volume,
-			Oscillator * _sub_osc ) :
-	m_waveShapeModel( _wave_shape_model ),
-	m_modulationAlgoModel( _mod_algo_model ),
-	m_freq( _freq ),
-	m_detuning( _detuning ),
-	m_volume( _volume ),
-	m_ext_phaseOffset( _phase_offset ),
-	m_subOsc( _sub_osc ),
-	m_phaseOffset( _phase_offset ),
-	m_phase( _phase_offset ),
-	m_userWave( NULL )
+Oscillator::Oscillator(const IntModel *wave_shape_model,
+			const IntModel *mod_algo_model,
+			const float &freq,
+			const float &detuning_div_samplerate,
+			const float &phase_offset,
+			const float &volume,
+			Oscillator *sub_osc) :
+	m_waveShapeModel(wave_shape_model),
+	m_modulationAlgoModel(mod_algo_model),
+	m_freq(freq),
+	m_detuning_div_samplerate(detuning_div_samplerate),
+	m_volume(volume),
+	m_ext_phaseOffset(phase_offset),
+	m_subOsc(sub_osc),
+	m_phaseOffset(phase_offset),
+	m_phase(phase_offset),
+	m_userWave(nullptr),
+	m_useWaveTable(false),
+	m_isModulator(false)
 {
 }
 
 
 
 
-void Oscillator::update( sampleFrame * _ab, const fpp_t _frames,
-							const ch_cnt_t _chnl )
+void Oscillator::update(sampleFrame* ab, const fpp_t frames, const ch_cnt_t chnl, bool modulator)
 {
-	if( m_freq >= Engine::mixer()->processingSampleRate() / 2 )
+	if (m_freq >= Engine::mixer()->processingSampleRate() / 2)
 	{
-		BufferManager::clear( _ab, _frames );
+		BufferManager::clear(ab, frames);
 		return;
 	}
-	if( m_subOsc != NULL )
+	// If this oscillator is used to PM or PF modulate another oscillator, take a note.
+	// The sampling functions will check this variable and avoid using band-limited
+	// wavetables, since they contain ringing that would lead to unexpected results.
+	m_isModulator = modulator;
```

Comment:
I made it a member variable because it was much simpler and a smaller change. To pass it as an argument to `getSample`, I would first need to propagate it from `update()` through each `updateXX()` and all the calls to different shape specializations. That's like 11 lines to change only for PM and there is 6 `updateXX()` variants in total, so 66 changes just to pass the parameter. Compared to one line.

Thinking back, I assume that may be also the reason why `m_freq` exists; if everything that defines a "global state" of the oscillator at the time of update was passed as an argument, the argument list would never end. Although I'm not entirely opposed to passing the `bool modulator` as a parameter if you think it is better or easier to follow.

Reply:
Hmm. Good point with the `m_freq` comparison, that might be the reason. In zyn, we just pass `m_freq` through all the sub functions, but the number of functions is much higher here due to template magic.

What about adding a comment above the `m_isModulator` delcaration, which states that it's only a class variable in order to have less function parameters, and that it can be easily converted into a function parameter if needed?

Reply:
Good idea with the comment; added.